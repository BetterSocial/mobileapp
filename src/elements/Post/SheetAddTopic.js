import React, {useEffect, useState} from 'react';
import {StyleSheet, Text, TouchableOpacity, View} from 'react-native';
import RBSheet from 'react-native-raw-bottom-sheet';
import {Button} from '../../components/Button';
import {colors} from '../../utils/colors';
import {fonts} from '../../utils/fonts';
import Icon from 'react-native-vector-icons/Fontisto';
import IconA5 from 'react-native-vector-icons/FontAwesome5';
import {TextInput} from 'react-native-gesture-handler';
import Gap from '../../components/Gap';
import TopicItem from '../../components/TopicItem';

const SheetAddTopic = ({refTopic, onAdd, topics}) => {
  const [dataTopic, setTopic] = useState('');
  const [listTopics, setlistTopics] = useState([]);
  const add = () => {
    if (dataTopic) {
      setlistTopics((val) => [...val, dataTopic]);
      setTopic('');
    }
  };
  const removeTopic = (v) => {
    let newArr = listTopics.filter((e) => e !== v);
    setlistTopics(newArr);
  };
  const merge = () => {
    setlistTopics(topics);
  };
  return (
    <RBSheet
      onOpen={() => merge()}
      ref={refTopic}
      closeOnDragDown={true}
      closeOnPressMask={true}
      customStyles={{
        container: {
          borderTopRightRadius: 20,
          borderTopLeftRadius: 20,
          height: 'auto',
        },
        draggableIcon: {
          backgroundColor: colors.alto,
        },
      }}>
      <View style={styles.container}>
        <Text style={styles.title}>Add topics</Text>
        <View style={styles.content}>
          <View style={styles.listItem}>
            {listTopics.map((value, index) => {
              return (
                <TopicItem
                  key={index}
                  label={value}
                  removeTopic={removeTopic}
                  style={{marginBottom: 12}}
                />
              );
            })}
          </View>
          {listTopics.length < 5 && (
            <View style={styles.containerInput}>
              <Text style={styles.hashtag}>#</Text>
              <TextInput
                style={{width: '100%', paddingStart: 0}}
                onSubmitEditing={() => add()}
                value={dataTopic}
                onChangeText={(v) => setTopic(v)}
                autoFocus={true}
                autoCapitalize="none"
                onKeyPress={({nativeEvent}) => {
                  if (nativeEvent.key.trim().length === 0) {
                    add();
                  }
                }}
              />
            </View>
          )}
        </View>
        {/* <Gap style={{height: 30}} /> */}
        <Text style={styles.textDesc}>
          Hit space to start a new topic. Add up to 5 topics.
        </Text>
        <Button onPress={() => onAdd(listTopics)}>Save</Button>
      </View>
    </RBSheet>
  );
};
export default SheetAddTopic;

const styles = StyleSheet.create({
  container: {
    paddingHorizontal: 20,
    paddingTop: 20,
    paddingBottom: 38,
  },
  content: {
    backgroundColor: colors.lightgrey,
    paddingHorizontal: 16,
    paddingTop: 17,
    // paddingBottom: 130,
    minHeight: 150,
    // height: 150,
    marginTop: 12,
    // marginBottom: 44,
    borderRadius: 8,
  },
  title: {
    color: colors.black,
    fontFamily: fonts.inter[600],
    fontSize: 18,
    fontWeight: 'bold',
  },
  containerTag: {
    backgroundColor: colors.white,
    borderRadius: 14,
    paddingHorizontal: 16,
    paddingVertical: 7.33,
    width: 122,
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    marginRight: 12,
    marginBottom: 12,
  },
  tag: {
    fontFamily: fonts.inter[400],
    fontSize: 12,
  },
  containerInput: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  listItem: {
    flexDirection: 'row',
    flexWrap: 'wrap',
  },
  hashtag: {
    fontSize: 12,
    fontWeight: '400',
    fontFamily: fonts.inter[400],
  },
  textDesc: {
    fontSize: 10,
    fontFamily: fonts.inter[400],
    color: colors.gray,
    marginTop: 5,
    marginBottom: 21,
  },
});
