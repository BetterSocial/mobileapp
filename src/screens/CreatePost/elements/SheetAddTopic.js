import * as React from 'react';
import {ScrollView, StyleSheet, Text, TextInput, TouchableNativeFeedback, View} from 'react-native';
import KeyEvent from 'react-native-keyevent';
import RBSheet from 'react-native-raw-bottom-sheet';

import {Button} from '../../../components/Button';
import TopicItem from '../../../components/TopicItem';
import {getTopics} from '../../../service/topics';
import {colors} from '../../../utils/colors';
import {fonts} from '../../../utils/fonts';
import {convertString} from '../../../utils/string/StringUtils';
import {isEmptyOrSpaces} from '../../../utils/Utils';
import Card from './Card';

const SheetAddTopic = ({refTopic, onAdd, topics, onClose, chatTopics}) => {
  const [dataTopic, setTopic] = React.useState('');
  const [listTopics, setlistTopics] = React.useState([]);
  const [chatTopic, setChatTopic] = React.useState([]);
  const [trigger, setTrigger] = React.useState(-1);
  const [topicSuggestion, setTopicSuggestion] = React.useState([]);
  const rbSheetHeight = 330;
  const [, setWidthInput] = React.useState(0);
  const inputRef = React.useRef();

  React.useEffect(() => {
    KeyEvent.onKeyUpListener((keyEvent) => {
      onKeyUp(keyEvent.keyCode);
    });

    return () => {
      KeyEvent.removeKeyUpListener();
    };
  }, []);

  React.useEffect(() => {
    add();
  }, [trigger]);

  const onKeyUp = (keycode) => {
    if (keycode === 62) {
      setTrigger(new Date().valueOf());
    }
  };

  const searchTopic = async (name) => {
    if (!isEmptyOrSpaces(name)) {
      getTopics(name)
        .then((v) => {
          if (v.data.length > 0) {
            if (v.data.length > 5) {
              const newData = v.data.slice(0, 5);
              setTopicSuggestion(newData);
            } else {
              setTopicSuggestion(v.data);
            }
          }
        })
        .catch((err) => {
          if (__DEV__) {
            console.log(err);
          }
        });
    }
  };
  const add = () => {
    const data = dataTopic.replace(/\s/g, '');
    if (data !== '' && !listTopics.includes(data)) {
      setlistTopics((val) => [...val, data]);
      setChatTopic((val) => [...val, `topic_${data}`]);
      setTopic('');
    }
    setTopic('');
    setTopicSuggestion([]);
  };
  const removeTopic = (v) => {
    const newArr = listTopics.filter((e) => e !== v);
    const newChatTopic = chatTopic.filter((chat) => chat !== `topic_${v}`);
    setlistTopics(newArr);
    setChatTopic(newChatTopic);
  };
  const merge = () => {
    setlistTopics(topics);
    setChatTopic(chatTopics);
  };
  const save = () => {
    const data = dataTopic.replace(/\s/g, '').toLowerCase();
    if (data === '') {
      onAdd(listTopics, chatTopic);
    } else if (!listTopics.includes(data)) {
      const newArr = [...listTopics, data];
      const newChatTopic = [...chatTopic, `topic_${data}`];
      onAdd(newArr, newChatTopic);
    }
    add();
    onClose();
  };
  // const onSwepDown = () => {
  //   const data = dataTopic.replace(/\s/g, '').toLowerCase();
  //   if (data === '') {
  //     saveOnClose(listTopics);
  //   } else if (!listTopics.includes(data)) {
  //       const newArr = [...listTopics, data];
  //       saveOnClose(newArr);
  //     }
  //   add();
  // };

  return (
    <RBSheet
      height={rbSheetHeight}
      onOpen={merge}
      // onClose={onSwepDown}
      ref={refTopic}
      closeOnDragDown={true}
      closeOnPressMask={true}
      customStyles={{
        container: styles.containerSheet,
        draggableIcon: styles.draggableIcon
      }}>
      <View style={styles.container}>
        <ScrollView showsVerticalScrollIndicator={false} keyboardShouldPersistTaps="always">
          <Text style={styles.title}>Add topics</Text>
          <View
            style={styles.content}
            onLayout={(event) => {
              // const headerHeightLayout = event.nativeEvent.layout.height;
              const headerWidthLayout = event.nativeEvent.layout.width;
              setWidthInput(headerWidthLayout);
            }}>
            <View style={styles.listItem}>
              {listTopics.map((value, index) => (
                <TopicItem
                  key={index}
                  label={value}
                  removeTopic={removeTopic}
                  style={styles.topicItem}
                />
              ))}
            </View>
            {listTopics.length < 5 && (
              <View style={styles.containerInput}>
                <Text style={styles.hashtag}>#</Text>
                <TextInput
                  ref={inputRef}
                  style={styles.input}
                  onSubmitEditing={() => add()}
                  value={dataTopic}
                  onChangeText={(v) => {
                    if (v.match(/\s+$/gm)) {
                      return add();
                    }
                    setTopic(v);
                    if (v !== '') {
                      searchTopic(v);
                    }

                    return null;
                  }}
                  autoCapitalize="none"
                  blurOnSubmit={false}
                  autoFocus
                />
              </View>
            )}
            {topicSuggestion.length > 0 && (
              <Card>
                {topicSuggestion.map((item, index) => (
                  <TouchableNativeFeedback
                    key={`topicSuggestions-${index}`}
                    onPress={() => {
                      let textTopic = convertString(item.name, ' ', '');
                      textTopic += ' ';
                      setTopic(textTopic.toLowerCase());
                      setTopicSuggestion([]);
                      onKeyUp(62);
                    }}>
                    <View style={{marginBottom: 5}}>
                      <Text
                        style={{
                          color: '#000000',
                          fontFamily: fonts.inter[500],
                          fontWeight: '500',
                          fontSize: 12,
                          lineHeight: 18
                        }}>
                        #{convertString(item.name, ' ', '')}
                      </Text>
                      {index !== topicSuggestion.length - 1 && (
                        <View style={{height: 1, marginTop: 5, backgroundColor: '#C4C4C4'}} />
                      )}
                    </View>
                  </TouchableNativeFeedback>
                ))}
              </Card>
            )}
          </View>

          <Text style={styles.textDesc}>Hit space to start a new topic. Add up to 5 topics.</Text>
          <Button onPress={() => save()}>Save</Button>
        </ScrollView>
      </View>
    </RBSheet>
  );
};
export default SheetAddTopic;

const styles = StyleSheet.create({
  container: {
    paddingHorizontal: 20,
    paddingTop: 20,
    paddingBottom: 38
  },
  content: {
    backgroundColor: colors.lightgrey,
    paddingHorizontal: 16,
    paddingTop: 17,
    minHeight: 150,
    marginTop: 12,
    borderRadius: 8,
    paddingBottom: 16
  },
  title: {
    color: colors.black,
    fontFamily: fonts.inter[600],
    fontSize: 18,
    fontWeight: 'bold'
  },
  containerTag: {
    backgroundColor: colors.white,
    borderRadius: 14,
    paddingHorizontal: 16,
    paddingVertical: 7.33,
    width: 122,
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    marginRight: 12,
    marginBottom: 12
  },
  tag: {
    fontFamily: fonts.inter[400],
    fontSize: 12
  },
  containerInput: {
    flexDirection: 'row',
    alignItems: 'center'
  },
  listItem: {
    flexDirection: 'row',
    flexWrap: 'wrap'
  },
  hashtag: {
    fontSize: 14,
    fontWeight: '400',
    fontFamily: fonts.inter[400]
    // marginTop: Platform.OS === 'android' ? -8 : 0,
  },
  textDesc: {
    fontSize: 10,
    fontFamily: fonts.inter[400],
    color: colors.gray,
    marginTop: 5,
    marginBottom: 21
  },
  containerSheet: {
    borderTopRightRadius: 20,
    borderTopLeftRadius: 20
  },
  draggableIcon: {
    backgroundColor: colors.alto
  },
  topicItem: {
    marginBottom: 12
  },
  input: {width: '100%', paddingStart: 0}
});
