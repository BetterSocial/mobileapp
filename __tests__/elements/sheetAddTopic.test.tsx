import * as React from 'react';
import {shallow, configure} from 'enzyme';
import toJson from 'enzyme-to-json';
import renderer from 'react-test-renderer';
import Adapter from 'enzyme-adapter-react-16';
import SheetAddTopic from '../../src/elements/Post/SheetAddTopic';

configure({adapter: new Adapter()});
jest.useFakeTimers();

describe('component SheetAddTopic', () => {
  let data = ['corona', 'indonesia'];
  it('SheetAddTopic snapshot', () => {
    const myRef = {
      current: {
        open: () => {},
        close: () => {},
      },
    };
    const component = shallow(
      <SheetAddTopic
        refTopic={myRef}
        onAdd={() => {}}
        topics={data}
        onClose={() => {}}
        saveOnClose={() => {}}
      />,
    );
    expect(toJson(component)).toMatchSnapshot();
  });

  it('SheetAddTopic renders correctly', () => {
    const myRef = {
      current: {
        open: () => {},
        close: () => {},
      },
    };

    const tree = renderer
      .create(
        <SheetAddTopic
          refTopic={myRef}
          onAdd={() => {}}
          topics={data}
          onClose={() => {}}
          saveOnClose={() => {}}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
