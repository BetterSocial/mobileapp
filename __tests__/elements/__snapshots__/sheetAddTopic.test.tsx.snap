// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component SheetAddTopic SheetAddTopic renders correctly 1`] = `
<Modal
  animationType="none"
  hardwareAccelerated={false}
  onRequestClose={[Function]}
  supportedOrientations={
    Array [
      "portrait",
      "portrait-upside-down",
      "landscape",
      "landscape-left",
      "landscape-right",
    ]
  }
  transparent={true}
  visible={false}
>
  <View
    onLayout={[Function]}
    style={
      Array [
        Array [
          Object {
            "backgroundColor": "#00000077",
            "flex": 1,
          },
          undefined,
        ],
        Object {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "transparent",
          "flex": 1,
          "opacity": 1,
        }
      }
    />
    <View
      onMoveShouldSetResponder={[Function]}
      onMoveShouldSetResponderCapture={[Function]}
      onResponderEnd={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderReject={[Function]}
      onResponderRelease={[Function]}
      onResponderStart={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      onStartShouldSetResponderCapture={[Function]}
      style={
        Object {
          "backgroundColor": "#fff",
          "borderTopLeftRadius": 20,
          "borderTopRightRadius": 20,
          "height": "auto",
          "overflow": "hidden",
          "transform": Array [
            Object {
              "translateX": 0,
            },
            Object {
              "translateY": 0,
            },
          ],
          "width": "100%",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "transparent",
            "width": "100%",
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "backgroundColor": "#ccc",
                "borderRadius": 5,
                "height": 5,
                "margin": 10,
                "width": 35,
              },
              Object {
                "backgroundColor": "#E0E0E0",
              },
            ]
          }
        />
      </View>
      <View
        style={
          Object {
            "paddingBottom": 38,
            "paddingHorizontal": 20,
            "paddingTop": 20,
          }
        }
      >
        <RCTScrollView
          keyboardShouldPersistTaps="always"
          showsVerticalScrollIndicator={false}
        >
          <View>
            <Text
              style={
                Object {
                  "color": "#000000",
                  "fontFamily": "Inter-Semi-bold",
                  "fontSize": 18,
                  "fontWeight": "bold",
                }
              }
            >
              Add topics
            </Text>
            <View
              style={
                Object {
                  "backgroundColor": "#F5F5F5",
                  "borderRadius": 8,
                  "marginTop": 12,
                  "minHeight": 150,
                  "paddingHorizontal": 16,
                  "paddingTop": 17,
                }
              }
            >
              <View
                style={
                  Object {
                    "flexDirection": "row",
                    "flexWrap": "wrap",
                  }
                }
              />
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                  }
                }
              >
                <Text
                  style={
                    Object {
                      "fontFamily": "Inter-Regular",
                      "fontSize": 12,
                      "fontWeight": "400",
                    }
                  }
                >
                  #
                </Text>
                <TextInput
                  allowFontScaling={true}
                  autoCapitalize="none"
                  blurOnSubmit={false}
                  collapsable={false}
                  onChangeText={[Function]}
                  onGestureHandlerEvent={[Function]}
                  onGestureHandlerStateChange={[Function]}
                  onKeyPress={[Function]}
                  onSubmitEditing={[Function]}
                  rejectResponderTermination={true}
                  style={
                    Object {
                      "paddingStart": 0,
                      "width": "100%",
                    }
                  }
                  underlineColorAndroid="transparent"
                  value=""
                />
              </View>
            </View>
            <Text
              style={
                Object {
                  "color": "#828282",
                  "fontFamily": "Inter-Regular",
                  "fontSize": 10,
                  "marginBottom": 21,
                  "marginTop": 5,
                }
              }
            >
              Hit space to start a new topic. Add up to 5 topics.
            </Text>
            <View
              accessible={true}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "backgroundColor": "#00ADB5",
                  "borderRadius": 8,
                  "flexDirection": "row",
                  "height": 48,
                  "justifyContent": "center",
                  "paddingHorizontal": 25,
                  "paddingVertical": 12,
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "#FFFFFF",
                    "fontFamily": "Inter-Semi-bold",
                    "fontSize": 18,
                  }
                }
              >
                Save
              </Text>
            </View>
          </View>
        </RCTScrollView>
      </View>
    </View>
  </View>
</Modal>
`;

exports[`component SheetAddTopic SheetAddTopic snapshot 1`] = `
<RBSheet
  animationType="none"
  closeDuration={200}
  closeOnDragDown={true}
  closeOnPressBack={true}
  closeOnPressMask={true}
  customStyles={
    Object {
      "container": Object {
        "borderTopLeftRadius": 20,
        "borderTopRightRadius": 20,
        "height": "auto",
      },
      "draggableIcon": Object {
        "backgroundColor": "#E0E0E0",
      },
    }
  }
  dragFromTopOnly={false}
  height={260}
  keyboardAvoidingViewEnabled={false}
  minClosingHeight={0}
  onClose={[Function]}
  onOpen={[Function]}
  openDuration={300}
>
  <View
    style={
      Object {
        "paddingBottom": 38,
        "paddingHorizontal": 20,
        "paddingTop": 20,
      }
    }
  >
    <ScrollView
      keyboardShouldPersistTaps="always"
      showsVerticalScrollIndicator={false}
    >
      <Text
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Inter-Semi-bold",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        Add topics
      </Text>
      <View
        style={
          Object {
            "backgroundColor": "#F5F5F5",
            "borderRadius": 8,
            "marginTop": 12,
            "minHeight": 150,
            "paddingHorizontal": 16,
            "paddingTop": 17,
          }
        }
      >
        <View
          style={
            Object {
              "flexDirection": "row",
              "flexWrap": "wrap",
            }
          }
        />
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
            }
          }
        >
          <Text
            style={
              Object {
                "fontFamily": "Inter-Regular",
                "fontSize": 12,
                "fontWeight": "400",
              }
            }
          >
            #
          </Text>
          <Component
            autoCapitalize="none"
            blurOnSubmit={false}
            onChangeText={[Function]}
            onKeyPress={[Function]}
            onSubmitEditing={[Function]}
            style={
              Object {
                "paddingStart": 0,
                "width": "100%",
              }
            }
            value=""
          />
        </View>
      </View>
      <Text
        style={
          Object {
            "color": "#828282",
            "fontFamily": "Inter-Regular",
            "fontSize": 10,
            "marginBottom": 21,
            "marginTop": 5,
          }
        }
      >
        Hit space to start a new topic. Add up to 5 topics.
      </Text>
      <Btn
        onPress={[Function]}
      >
        Save
      </Btn>
    </ScrollView>
  </View>
</RBSheet>
`;
